Improving performance for ScrollingObjectCollection

MaxWang-MS opened the pull request:
## Overview
This PR is ported from #9144 to allow direct edits by MRTK maintainers.

The PR modifies Clipping Primitive, Material Instance and Scrolling Object Collection scripts to decrease allocs and frame time. Part of the improvement is transparent to user, and the other part requires a manual opt-in in Clipping Primitive/Material Instance. Some other modifications in #9144 are left out due to performance and other considerations after testing on HoloLens 2.

Big thanks to @Agredek for making the contribution! The original notes of the PR are attached below:

> ## Overview
> 
> Performance improvements for CoreServices, ClippingPrimitive and, mostly, for ScrollingObjectCollection.
> 
> ## Changes
> - Mixed Reality Services are now correctly cached inside CoreServices.cs
> - ClippingPrimitive now caches Renderers and MaterialInstances instead of asking for them in every frame
> - MaterialInstance now caches shared materials
> - ScrollingObjectCollection uses ScrollingObjectCollectionContainer to find and work on child Renderers and Colliders
> - ScrollingObjectCollectionContainer is responsible for finding and storing references for it's children's renderers and colliders. It creates a small overhead whenever children are added or removed, but vastly enhances performance for each ScrollingObjectCollection Update call.
> 
> 
> ## Verification
> Hi! My name's Daniel and I work for apoQlar GmbH. During one of our technical sprints, we decided to work on performance improvements. Changes that we have introduced have helped us to achieve 60 fps and vastly mitigate GC impact. We hope that you'll find those changes suitable.

keveleigh commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -16,10 +17,18 @@ namespace Microsoft.MixedReality.Toolkit.Utilities
-        protected List<Renderer> renderers = new List<Renderer>();
-
+        private List<Renderer> renderersCache = new List<Renderer>(InitialCollectionSize);
+        
```
```suggestion

```

keveleigh commented on Assets/MRTK/SDK/Features/UX/Scripts/ScrollingObjectCollection/ScrollingObjectCollection.cs:
```diff
@@ -16,14 +15,14 @@ namespace Microsoft.MixedReality.Toolkit.UI
     /// A scrollable frame where content scroll is triggered by manual controller click and drag or according to pagination settings.
-    //// </summary>
+    /// </summary>
     ///<remarks>Executing also in edit mode to properly catch and mask any new content added to scroll container.</remarks>
```
```suggestion
    /// <remarks>Executing also in edit mode to properly catch and mask any new content added to scroll container.</remarks>
```

keveleigh commented on Assets/MRTK/SDK/Features/UX/Scripts/ScrollingObjectCollection/ScrollingObjectCollectionContainer.cs:
```diff
@@ -0,0 +1,97 @@
+{
+    /// <summary>
+    /// Container for all renderers and colliders used in the ScrollingObjectCollection.
+    /// It's main purpose is to find and cache those Components.
```
```suggestion
    /// Its main purpose is to find and cache those Components.
```

keveleigh commented:
/azp run

MaxWang-MS commented:
As discussed with @davidkline-ms earlier today, the proposed modification to core services will introduce references to the service instances, which would prevent the GC from happening when a service is destroyed. That is why the TestDynamicServices test is failing. We need to consider whether we should go down this route or not.

Cameron-Micka commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -16,10 +17,17 @@ namespace Microsoft.MixedReality.Toolkit.Utilities
     [HelpURL("https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Rendering/ClippingPrimitive.html")]
     public abstract class ClippingPrimitive : MonoBehaviour, IMaterialInstanceOwner
     {
+        private const int InitialCollectionSize = 256;
```
Might be good to sanity check how big a 256 member Dictionary is. Might be a bit wasteful for most use cases? But might be fine since these are all reference types.

Agredek replied:
This value has been chosen with a reserve in mind (around 2-times). Depending on the button visuals complexity level it should be enough for around 10 buttons. In our application, lists are currently limited to 5 buttons. The most common size of the collection is between 80-100 items. So yes - it is a little bit wasteful, although while deciding the initial size of the collection I didn't want to only adhere to our solution.

Cameron-Micka commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -127,57 +136,51 @@ public bool IsDirty
-        /// <param name="_renderer">The renderer to add.</param>
-        public void AddRenderer(Renderer _renderer)
+        /// <param name="renderer">The renderer to add.</param>
+        public void AddRenderer(Renderer renderer)
```
Is the member name "renderer" allowed now? It use to cause a collision with Unity's Component type.

MaxWang-MS replied:
I think so as there is no complain from VS or during build time and also there are other places where `renderer` is used as a variable name.

Cameron-Micka commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -127,57 +136,51 @@ public bool IsDirty
-                    renderers.Add(_renderer);
-                }
+                    var materialInstance = renderer.EnsureComponent<MaterialInstance>();
+                    if (!materialInstance)
```
I don't think EnsureComponent will return null? Unless AddComponent fails.

Agredek replied:
It's just a sanity check. As you said - It can return null if AddComponent fails. It may fail if gameObject has already been destroyed.

Cameron-Micka commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -188,12 +191,15 @@ private void RemoveRenderer(int index)
         public void ClearRenderers()
         {
-            if (renderers != null)
+            if (renderers.Count > 0)
```
I have a faint memory of renderers being null is some editor scenarios. Unity may have addressed this since? But, work testing.

Agredek replied:
Renderers collection is now readonly. In my testing I have never seen it being null.

Cameron-Micka commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -16,10 +17,17 @@ namespace Microsoft.MixedReality.Toolkit.Utilities
+        private const int InitialCollectionSize = 256;
+
+        /// <summary>
+        /// The renderer(s) that should be affected by the primitive. This collection is a copy of internal HashSet.
```
nit: little misleading since this is not a HashSet, it's a Dictionary.

Agredek replied:
Indeed, it's my slip-up 😅 I didn't notice.

Cameron-Micka reviewed (changes requested):
Thanks for these changes! I have a few concerns over about the MaterialInstance changes that I commented on.

MaxWang-MS commented:
@Agredek Please take a look at the comments above when you are available. Also I noticed the example scene at Assets/MRTK/Examples/Demos/ScrollingObjectCollection/Scenes/ScrollingObjectCollection.unity is not working properly in this PR. Could you double check that as well? Thanks!

Agredek commented:
> @Agredek Please take a look at the comments above when you are available. Also I noticed the example scene at Assets/MRTK/Examples/Demos/ScrollingObjectCollection/Scenes/ScrollingObjectCollection.unity is not working properly in this PR. Could you double check that as well? Thanks!

@MaxWang-MS I have checked out the example scene. Sorry, I totally forgot to check them out while working on those changes. Indeed, this scene doesn't work correctly because current implementation of ScrollingObjectCollectionContainer doesn't take into consideration pre-populated collections. 

Luckliy, this change is really simple and consists only of calling OnTransformChildrenChanged() in Awake/Start method in ScrollingObjectCollectionContainer class. Is there some way for me to commit those changes?

Agredek commented on Assets/MRTK/Core/Utilities/Rendering/MaterialInstance.cs:
```diff
@@ -146,7 +155,7 @@ private void Awake()
         {
             // If the materials get changed via outside of MaterialInstance.
-            var sharedMaterials = CachedRenderer.sharedMaterials;
+            var sharedMaterials = CachedSharedMaterials;
```
Not being part of the MRTK Team, I don't really understand why would we want to check whether materials has been changed outside of the MaterialInstance API. If this is Editor-only, then maybe this code should be sorrounded by #if preprocessor directive?

Cameron-Micka replied:
This is a great question. Ideally we wouldn't worry about developers changing materials outside of the MaterialInstance API. But, I worried that some developers wouldn't be aware of the existence of the MaterialInstance and blindly assign to the renderer's shared material (causing the whole system to fall apart). 

As you point out this workflow is unavoidable in the editor, so maybe it's best to just wrap this in an editor preprocessor and fail when a developer assigns to the renderer's shared materials at runtime? @MaxWang-MS do you have any thoughts? We could also wrap this check in a `DEBUG` preprocessor so it runs in editor and in development builds and runs unchecked in release?

MaxWang-MS replied:
I think it is likely that someone not very familiar with MRTK will make the mistake of modifying shared material directly instead of going through the material instance class, so I think we should either keep we what we had before or add a default-to-false switch that enables this optimization. I am leaning towards the former unless calling render.sharedMaterials is actually expensive.

Cameron-Micka replied:
I agree with the former since much of the MRTK still doesn't track instance materials appropriately. Maybe we can make this change after #6166 is addressed?

Agredek replied:
Just to put my two cents: iirc, after replacing renderer.sharedMaterials calls with this cache we have gained around 0.5ms and removed some garbage allocation. As you know, in AR environment it's a lot. Renderer.sharedMaterials useses some internal extern method that caught my attention while analyzing the Profiler.

Cameron-Micka replied:
I agree anytime we reach into a container held by the engine we incur a bunch of overhead moving and copying from native to managed. I hate to complicate this, but maybe we make it an "opt-in" optimization for users that need it? Else, the behavior is the same for existing users?

MaxWang-MS replied:
Hi @Cameron-Micka, I made this optimization opt-in as you suggested. Please let me know if it looks good now.

MaxWang-MS commented:
/azp run

MaxWang-MS commented:
> > @Agredek Please take a look at the comments above when you are available. Also I noticed the example scene at Assets/MRTK/Examples/Demos/ScrollingObjectCollection/Scenes/ScrollingObjectCollection.unity is not working properly in this PR. Could you double check that as well? Thanks!
> 
> @MaxWang-MS I have checked out the example scene. Sorry, I totally forgot to check them out while working on those changes. Indeed, this scene doesn't work correctly because current implementation of ScrollingObjectCollectionContainer doesn't take into consideration pre-populated collections.
> 
> Luckliy, this change is really simple and consists only of calling OnTransformChildrenChanged() in Awake/Start method in ScrollingObjectCollectionContainer class. Is there some way for me to commit those changes?

@Agredek Thanks for the reply! I have updated the code accordingly. Also we had an internal discussion and decided to not modify CoreServices at the moment due to the possibilities of that change breaks existing projects. We are planning to revisit that change when moving onto the next major release. I also made some minor modifications so that now most tests are passing. Could you take a look at the `ChildrenCanBeAddedAndDeleted` test in ScrollViewTests.cs? That one seems to be still failing and I was not able to find out the exact cause. During debugging I believe I observed that the clipping is not working the same way as before but I can be wrong.

Agredek commented:
> @Agredek Thanks for the reply! I have updated the code accordingly. Also we had an internal discussion and decided to not modify CoreServices at the moment due to the possibilities of that change breaks existing projects. We are planning to revisit that change when moving onto the next major release. I also made some minor modifications so that now most tests are passing. Could you take a look at the `ChildrenCanBeAddedAndDeleted` test in ScrollViewTests.cs? That one seems to be still failing and I was not able to find out the exact cause. During debugging I believe I observed that the clipping is not working the same way as before but I can be wrong. Also I sent an invite to you so that you can become a collaborator of my fork and thus gain write access to this branch (MaxWang-MS:improve_perf).

Hi! Sorry for such a long delay. We had a lot of internal work and only now I have found some time to check out the tests you've been talking about.
Indeed, there was a problem with ChildrenCanBeAddedAndDeleted test. It seems that problems lied with my changes (or lack of them) in ScrollingObjectCollection as well as in ScrollingObjectCollectionContainer. Container wasn't notified about the changes made to the transform if children were deeper into the GameObjects tree. Also removing items from the ScrollingObjectCollection wasn't using my API so I fixed that as well.
Tests are running smoothly now. Hope that I haven't made any more mistakes ;)

MaxWang-MS commented:
/azp run

MaxWang-MS commented:
> > @Agredek Thanks for the reply! I have updated the code accordingly. Also we had an internal discussion and decided to not modify CoreServices at the moment due to the possibilities of that change breaks existing projects. We are planning to revisit that change when moving onto the next major release. I also made some minor modifications so that now most tests are passing. Could you take a look at the `ChildrenCanBeAddedAndDeleted` test in ScrollViewTests.cs? That one seems to be still failing and I was not able to find out the exact cause. During debugging I believe I observed that the clipping is not working the same way as before but I can be wrong. Also I sent an invite to you so that you can become a collaborator of my fork and thus gain write access to this branch (MaxWang-MS:improve_perf).
> 
> @MaxWang-MS Hi! Sorry for such a long delay. We had a lot of internal work and only now I have found some time to check out the tests you've been talking about.
> Indeed, there was a problem with ChildrenCanBeAddedAndDeleted test. It seems that problems lied with my changes (or lack of them) in ScrollingObjectCollection as well as in ScrollingObjectCollectionContainer. Container wasn't notified about the changes made to the transform if children were deeper into the GameObjects tree. Also removing items from the ScrollingObjectCollection wasn't using my API so I fixed that as well.
> Tests are running smoothly now. Hope that I haven't made any more mistakes ;)

No worries! I have been distracted by some other work lately as well.
For the material instance class, I agree with @Cameron-Micka that we probably should make the optimization optional so that things do not break for existing users.
For the fix for the test failure, I see you dealing with the issue where the children is one-level deeper. But it does not solve the case where the children are even deeper. I thought about an alternative where the Children array in the container class stores references to each of the content/item in the ScrollingObjectCollection(SOC) instead of the first-level children now (often it is a single GameObject with the GridObjectCollection (GOC)). The benefit of this approach is that when removing objects from SOC instead of iterating through the keys of ItemRenderersMap and manually performing checks on each renderer we can simply use ItemRenderersMap as a dictionary to get the renderers we need. The complication is that it would be hard to determine the "real" content/item gameObject of SOC as people do not always use GOC. So it could be possible that the first-level children of SOC are the "real" content/items or it could also be possible that those "real" stuff are second-level children of SOC (going one level deeper), and that is not considering some special customized implementation where things are even deeper.

mihaits commented:
are there any updates on this?

mihaits commented:
I tried including these changes in the project I'm working on because I had performance problems with the expected number of buttons in a list. I've also included the changes since this PR until the 2.7.3 release. I'm documenting the bugs I found in case this PR is still considered:
- ClippingPrimitive doesn't work with renderers pre-assigned in the editor, only with AddRenderer.
- ClippingPrimitive functionality is broken with TextMeshPro if the TMP component or its GameObject is disabled and enabled again. This happens because of the MaterialInstance changes.
- the previous bug breaks text clipping when "Disable Clipped Game Objects" is enabled on ScrollingObjectCollection, limiting its use to "Disable Clipped Renderers". With enough buttons this becomes a performance problem when the Interactable.Update calls start adding up

As a result I've reverted the changes and settled on using simpler buttons or a custom paginated list with item object pooling in the future.
Unity 2021.2.8 + MRTK 2.7.3

MaxWang-MS commented:
/azp run

MaxWang-MS commented:
/azp run

MaxWang-MS commented:
/azp run

Cameron-Micka reviewed (approved):
Thank you for updating this PR!

Did a visual inspection of the changes. But would be good to test in editor/player as well. 😊

MaxWang-MS commented:
/azp run

MaxWang-MS commented:
Hi @mihaits, could you check if the latest commits solve the issues you were having?

mihaits commented:
hi @MaxWang-MS, thanks for the updates! I will test these changes as soon as I can get to them in the following week.

keveleigh commented on Assets/MRTK/SDK/Features/UX/Scripts/ScrollingObjectCollection/ScrollingObjectCollection.cs:
```diff
@@ -1476,10 +1479,10 @@ private void HandleFalloffPerFrame()
                         }
                         else
                         {
-                            avgVelocity *= velocityDampen;
```
nit, but seems like some whitespaces changes here that could be reverted

MaxWang-MS replied:
Fixed!

keveleigh commented on Assets/MRTK/SDK/Features/UX/Scripts/ScrollingObjectCollection/ScrollingObjectCollection.cs:
```diff
@@ -567,7 +566,7 @@ public int FirstHiddenCellIndex
             }
         }
-
+        
```
Other whitespace. Maybe worth formatting this file in VS?

MaxWang-MS replied:
Fixed!

keveleigh commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -339,18 +397,14 @@ protected virtual void UpdateRenderers()
-            if (renderers != null)
-            {
-                for (var i = 0; i < renderers.Count; ++i)
+            if (renderers.Count > 0)
```
nit, but this `if` looks like it's missing its braces

MaxWang-MS replied:
Fixed!

keveleigh commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -304,13 +363,12 @@ protected virtual void Initialize()
         protected virtual void UpdateRenderers()
         {
-            if (renderers == null) { return; }
+            if (renderers.Count == 0) { return; }
```
Are we guaranteed that `renderers` will never be null?

MaxWang-MS replied:
I think so, but added back the check just in case.

MaxWang-MS commented:
/azp run

keveleigh commented on Assets/MRTK/Core/Utilities/StandardShader/ClippingPrimitive.cs:
```diff
@@ -206,11 +239,37 @@ public void ClearRenderers(bool autoDestroyMaterial = true)
+                    }
+                }
+            }
+            if (renderers.Count != rendererDict.Count)
```
Is this case ever possible to hit? It feels like it should be accounted for in the above `if`, where we populate the dictionary directly from the list.

MaxWang-MS replied:
I put it here in case there are duplicates in `renderers`. It is theoretically possible as `renderers` is a list. I don't really expect this to happen when people know what they are supposed to do but still put it here to make sure the list and the dict are always in sync after `Awake()`.
